### ISSUE TITLE
Document why `DYNAMIC_GHC_PROGRAMS=NO` is used for HLS GHC bindists

### ISSUE THREAD

**Issue by nh2** [haskell/haskell-language-server#2000]
I learned from @mpickering that HLS distributes GHC binaries with `DYNAMIC_GHC_PROGRAMS=YES`, and that this creates various problems for TemplateHaskell regarding dynamic linking.

I am interested in this topic because of https://github.com/NixOS/nixpkgs/issues/129245.

I was trying to find documentation on the fact that, or why, HLS does that, but couldn't find any.

Could that be written down somewhere? It would make users' and packagers' life easier.

---

On the question itself: Would `DYNAMIC_GHC_PROGRAMS=YES` be more problematic to distribute somwhow? It appears to me that if e.g. the VSCode plugin downloader can download GHC, it could also download some `.so` files with it.

**Comment by pepeiborra** [comment]
There seems to be some confusion here. `DYNAMIC_GHC_PROGRAMS` is a ghc flag and has nothing to do with HLS. 

I think you are referring to the fact that distribution HLS binaries are linked statically, that is, both Haskell packages and external libraries (modulo glibc) are statically linked in a single binary. This is done for ease of distribution. 

The problem comes because the ghc package is linked statically, at runtime the RTS linker will be used to dynamically load packages and their external dependencies, as documented [here](https://gitlab.haskell.org/ghc/ghc/-/wikis/dynamic-linking-debate#do-we-need-the-static-linker-anyway). Unfortunately the RTS linker is far from perfect and has issues in all platforms. 

GHC binaries switched to `DYNAMIC_GHC_PROGRAMS=YES` in 7.8/7.10, except on Windows where I think they are still statically linked. This was done for good reason: TH and GHCi are able to use `dlopen` to load packages and libraries. 

Ideally, the choice of linker/loader would be decoupled from whether the GHC package is linked statically or dynamically. I remember asking @bgamari at some point to write down what needs to be done in a ticket for future reference.

**Comment by nh2** [comment]
> There seems to be some confusion here. `DYNAMIC_GHC_PROGRAMS` is a ghc flag and has nothing to do with HLS.
> 
> I think you are referring to the fact that distribution HLS binaries are linked statically

Yes, that's what I meant with _"HLS distributes GHC binaries with `DYNAMIC_GHC_PROGRAMS=YES`_".

> as documented [here](https://gitlab.haskell.org/ghc/ghc/-/wikis/dynamic-linking-debate#do-we-need-the-static-linker-anyway)

Thanks for linking me that page. It occurred to me now that that the _"What happens if we build a static executable and link in the GHC package?"_ mentioned in there also applies to the `ghc` binary itself.

> This is done for ease of distribution.

Could you elaborate in which way it makes distribution easier? Is it just have less files around, or are there othere benefits?

Thanks!

**Comment by pepeiborra** [comment]
> > There seems to be some confusion here. `DYNAMIC_GHC_PROGRAMS` is a ghc flag and has nothing to do with HLS.
> > I think you are referring to the fact that distribution HLS binaries are linked statically
> 
> Yes, that's what I meant with _"HLS distributes GHC binaries with `DYNAMIC_GHC_PROGRAMS=YES`_".
> 
> > as documented [here](https://gitlab.haskell.org/ghc/ghc/-/wikis/dynamic-linking-debate#do-we-need-the-static-linker-anyway)
> 
> Thanks for linking me that page. It occurred to me now that that the _"What happens if we build a static executable and link in the GHC package?"_ mentioned in there also applies to the `ghc` binary itself.
> 
> > This is done for ease of distribution.
> 
> Could you elaborate in which way it makes distribution easier? Is it just have less files around, or are there othere benefits?
> 
> Thanks!

Yes

**Comment by nh2** [comment]
Ah sorry, I just noticed what introduced confusion. I wrote:

> I learned from @mpickering that HLS distributes GHC binaries with `DYNAMIC_GHC_PROGRAMS=YES`,

but I meant to write:

> I learned from @mpickering that HLS distributes GHC binaries with `DYNAMIC_GHC_PROGRAMS=NO`

(that the GHC binaries distributed by HLS use `NO`).

If the main benefit to HLS of using `DYNAMIC_GHC_PROGRAMS=NO` is "less files", wouldn't the fact that TH does not work well with it be a big motivator to switch to `YES` instead, like the official GHC bindists?

More files seem like an OK tradeoff to make vs wrestling dynamic loader problems.

**Comment by pepeiborra** [comment]
I simply don't have the expertise to set up dynamically linked binaries for distribution (other than a simple Docker image). It would be awesome if someone else can make that happen.

**Comment by hasufell** [comment]
I consider this a GHC bug and I'd advise to pursue an upstream fix before going down the road of dynamic linking on linux. Static linking is a huge win for compatibility issues. Ultimately, we want GHC statically linked, so stack, ghcup etc don't have to guess which bindist is compatible with distro X version Y.

**Comment by adamse** [comment]
@hasufell which part do you consider a ghc bug?

**Comment by hasufell** [comment]
@adamse 

As @pepeiborra said

> Ideally, the choice of linker/loader would be decoupled from whether the GHC package is linked statically or dynamically. I remember asking @bgamari at some point to write down what needs to be done in a ticket for future reference.

**Comment by adamse** [comment]
Ah, that's a good point. I've thought about a little bit, and I think it should be possible :)

**Comment by adamse** [comment]
I've made a proof of concept at https://gitlab.haskell.org/ghc/ghc/-/commit/7e77131440115c10cf3b7d3b14b1cce57a5dc21f

build with

```
./hadrian/build --flavour=default+no_dynamic_ghc stage2:exe:ghc-bin stage1.ghc-bin.ghc.link+="-optl=-export-dynamic -debug" 
./hadrian/build --flavour=default+no_dynamic_ghc _build/ghc-stage2 stage1.ghc-bin.ghc.link+="-optl=-export-dynamic -debug" 
````

you should then be able to observe it loading .so's in the debug output, example command:

```
./_build/ghc-stage2 -e 'import Data.Text' -e 'import Data.Text.IO as T' -e 'T.putStrLn (pack "hej")'
```

**Comment by pepeiborra** [comment]
Right, so with this GHC patch, if we build HLS with `--export-dynamic`/`--dynamic-list`, then the static HLS binary will use the system linker to load THs?

**Comment by adamse** [comment]
Yes, I think that should work.

**Comment by hasufell** [comment]
Can someone backport that patch to 8.10.7 or 9.0.1 so we can test?

**Comment by adamse** [comment]
8.10: https://gitlab.haskell.org/ghc/ghc/-/commit/d8ad8b10d31d123869f36912d7f83822e287c990

**Comment by adamse** [comment]
For discussion how we proceed in GHC if it turns out to work for HLS: https://gitlab.haskell.org/ghc/ghc/-/issues/20628

**Comment by adamse** [comment]
do we have any reproducer for th issues that i can try this with?

**Comment by jneira** [comment]
@adamse many thanks for take care of this widespread issue. Revising the issues [labeled with template haskell related](https://github.com/haskell/haskell-language-server/issues?q=is%3Aopen+label%3A%22type%3A+template+haskell+related%22+repro) i've found some:
* https://github.com/haskell/haskell-language-server/issues/1982#issuecomment-868933053
* https://github.com/haskell/haskell-language-server/issues/1841 (point to a repro repo: https://github.com/FinleyMcIlwaine/hls-issue)
* https://github.com/haskell/haskell-language-server/issues/1297 (points to https://github.com/FinleyMcIlwaine/hls-issue)
* https://github.com/haskell/haskell-language-server/issues/469 (this one seemed to be macos specific, points to https://github.com/patrickt/possession)
* https://github.com/haskell/haskell-language-server/issues/800#issuecomment-589912942 (points to https://github.com/afcowie/hello/tree/ghcide-issue-444)
* https://github.com/haskell/haskell-language-server/issues/1737 (points to https://github.com/serokell/edna)

There will be some others in others th related issues. Sorry i cant verify all those reproducers are up to date and being correct.

**Comment by adamse** [comment]
I tried the reproducer from #1982 (https://github.com/aschmois/th-pq-segfault) with my latest GHC patch (https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7037).

I did not manage to build hls/ghcide with ghc head but I used ghci as a reasonable proxy for how hls uses the ghc api. I was able to successfully load the reproducer.

**Comment by pepeiborra** [comment]
Could you build static HLS 1.5.0 binaries for Mac OS with your patched GHC 8.10.7 and upload the binaries somewhere? Then @jneira can share them with users for testing.

**Comment by jneira** [comment]
nice!

> I used ghci as a reasonable proxy for how hls uses the ghc api

I am not sure 100% but iirc some of the issues were not reproduced in ghci. Did you check that ghci failed in that reproducer with a standard ghc? thanks!

**Comment by adamse** [comment]
yes, I managed to repro the issue (or an issue...) without my new flag: https://github.com/adamse/th-pq-segfault#readme

**Comment by adamse** [comment]
@pepeiborra I will build 8.10 and with HLS from that, but I can only do Linux.

**Comment by pepeiborra** [comment]
Nix ?

**Comment by adamse** [comment]
cross compiling is out of my league I'm afraid :)

**Comment by pepeiborra** [comment]
> cross compiling is out of my league I'm afraid :)

No, but if you have a Nix expression that builds hls with your patched GHC 8.10.7, I can run it locally or even install it in our CI

**Comment by adamse** [comment]
I see. I've lost most of my nix knowledge, but perhaps someone can whip something up from a ghc source dist? https://drive.google.com/file/d/1bO8G85u1IP8OUtqwBHV_aBx3dRA2WXOY/view?usp=sharing

**Comment by pepeiborra** [comment]
Invoking @berberman @maralorn @lazamar @domenkozar

**Comment by hasufell** [comment]
We don't need to cross-compile. Building in an alpine docker container is the easiest solution. But I wasn't able to build GHC with the linked patch. It failed to compile.

**Comment by adamse** [comment]
I built 8.10.7 and then ghcide, adding the `-optl=-export-dynamic` flag for linking. I end up with

```
ghcide version: 1.5.0 (GHC: 8.10) (PATH: /home/adam/src/ghc/ghc-8.10/hls/dist-newstyle/build/x86_64-linux/ghc-8.10.7/ghcide-1.5.0/x/ghcide/build/ghcide/ghcide) (GIT hash: 311107eabbf0537e0c192b2c377d282505b4eff1)
 ghcide setup tester in /home/adam/src/ghc/ghc-8.10/hls/th-pq-segfault.
Report bugs at https://github.com/haskell/haskell-language-server/issues

Step 1/4: Finding files to test in /home/adam/src/ghc/ghc-8.10/hls/th-pq-segfault
Found 3 files

Step 2/4: Looking for hie.yaml files that control setup
Found 1 cradle
  (/home/adam/src/ghc/ghc-8.10/hls/th-pq-segfault/hie.yaml)

Step 3/4: Initializing the IDE

Step 4/4: Type checking the files
[Info] Consulting the cradle for "src/B.hs"
[Info] Using interface files cache dir: /home/adam/.cache/ghcide/example-0.0.0-inplace-7ade6477eea70027383c9b80b9010ccbb3a93ff1
[Info] Making new HscEnv[example-0.0.0-inplace]
[Info] Consulting the cradle for "pqexample/src/PQ.hs"
[Info] Using interface files cache dir: /home/adam/.cache/ghcide/pqexample-0.0.0-inplace-6454242b35d045d040c1990c72fd4b7bb76a8951
[Info] Using interface files cache dir: /home/adam/.cache/ghcide/example-0.0.0-inplace-ff2a3ddd88ff9f698df1d4900670cb4f4469339b
[Info] Making new HscEnv[pqexample-0.0.0-inplace,example-0.0.0-inplace]
[Info] File:     /home/adam/src/ghc/ghc-8.10/hls/th-pq-segfault/src/B.hs
Hidden:   no
Range:    1:1-2:1
Source:   compiler
Severity: DsError
Message: 
  tried to look at linkable for GetModIfaceWithoutLinkable for NormalizedFilePath
  "/home/adam/src/ghc/ghc-8.10/hls/th-pq-segfault/pqexample/src/PQ.hs"
  CallStack (from HasCallStack):
  error, called at src/Development/IDE/Core/Rules.hs:888:79 in
  ghcide-1.5.0-inplace:Development.IDE.Core.Rules
[Info] finish: User TypeCheck (took 0.21s)
[Info] finish: GetHie (took 0.04s)
[Info] finish: GenerateCore (took 0.00s)
Files that failed:
 * /home/adam/src/ghc/ghc-8.10/hls/th-pq-segfault/src/B.hs
```

I've no idea what it means, is it success? If I do `ghcide typecheck src/B.hs` everything is fine.

**Comment by pepeiborra** [comment]
> I've no idea what it means, is it success? If I do ghcide typecheck src/B.hs everything is fine.

This is a regression in HLS 1.5, fixed in #2380. To work around it, you have 3 options:
1. typecheck only `src/B.hs` (`src/A.hs` is not enough to trigger the crash)
2. Apply the fix (#2380)
3. Test using HLS 1.4.0

**Comment by adamse** [comment]
Super, with that patch running ghcide on the th-pq-segfault example has no problems :)

**Comment by jneira** [comment]
maybe could we build the patch in https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7037 using ghcup?

**Comment by hasufell** [comment]
> maybe could we build the patch in https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7037 using ghcup?

ghc CI already builds GHC on every pipeline. So this commit https://gitlab.haskell.org/ghc/ghc/-/commit/d8ad8b10d31d123869f36912d7f83822e287c990 just needs to land in a proper branch and have CI triggered

**Comment by adamse** [comment]
most of the builds here were fine: https://gitlab.haskell.org/ghc/ghc/-/pipelines/44227, bindists available.

**Comment by jneira** [comment]
oh yeah, maybe would be too much CI abuse open a mr against the 9.0 branch? it would be good to have that change backported anyways

**Comment by adamse** [comment]
I made a backport to the 8.10 branch: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7039, hopefully CI will produce some bindists.

HLS changes required here: https://github.com/adamse/hls

**Comment by pepeiborra** [comment]
> HLS changes required here: https://github.com/adamse/hls

It might be worth submitting a PR using CPP to guard the change, like:

https://github.com/haskell/haskell-language-server/blob/d9b20ee5cf4482c94995bf7bdbd675677b602ec1/ghcide/src/Development/IDE/Core/Rules.hs#L1030-L1032

**Comment by jneira** [comment]
@adamse hi! not sure if you read the comment above about the CPP guard to use your patch (https://github.com/haskell/haskell-language-server/issues/2000#issuecomment-974843019)
any progress on th ghc side? i've just check https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7037 is not merged yet
please let us know if we could help in any way to get it merged
thanks!

**Comment by mpickering** [comment]
@jneira The patch isn't merged because it doesn't appear to be functionally correct (ie the testsuite the doesn't pass).
